service/history/workflow_task_handler.go

```diff
@@ -211,6 +213,58 @@ func (handler *workflowTaskHandlerImpl) handleCommands(
 	return mutations, nil
 }
 
+func (handler *workflowTaskHandlerImpl) rejectUnprocessedUpdates(
+	ctx context.Context,
+	workflowTaskScheduledEventID int64,
+	workflowTaskHeartbeating bool,
+	wfKey definition.WorkflowKey,
+	workerIdentity string,
+) error {
+
+	// If server decided to fail WT (instead of completing), don't reject updates.
+	// New WT will be created, and it will deliver these updates again to the worker.
+	// Worker will do full history replay, and updates should be delivered again.
+	if handler.workflowTaskFailedCause != nil {
+		return nil
+	}
+
+	// If WT is a heartbeat WT, then it doesn't have to have messages.
+	// TODO (alex-update): Don't deliver these updates on the next WT.
+	//  Fix TestUpdateWorkflow_SpeculativeWorkflowTask_Heartbeat test.
+	if workflowTaskHeartbeating {
+		return nil
+	}
+
+	// If worker has just completed workflow with one of the WF completion command,
+	// then it might skip processing some updates. In this case, it doesn't indicate old SDK or bug,
+	// but it's a normal behavior. Updates will be rejected with "workflow is closing" reason though.
```

Comment:
>Updates will be rejected with "workflow is closing" reason though

If an update is accepted, but not complete. What will happen? I'd argue it should fail with the "workflow is closing" reason.

Reply:
I changed the comment. It now says "all unprocessed updates". I need to think what to do with accepted but not completed updates. There will be another PR for this.